oneflow/core/autograd/autograd_engine.cpp

```
@@ -0,0 +1,129 @@
+/*
+Copyright 2020 The OneFlow Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "oneflow/core/autograd/autograd_engine.h"
+#include "oneflow/core/framework/tensor.h"
+#include "oneflow/core/framework/tensor_arg.h"
+#include "oneflow/core/framework/tensor_tuple.h"
+
+namespace oneflow {
+namespace one {
+
+namespace {
+
+bool IsReadyToRun(const std::vector<std::shared_ptr<TensorArg>>& out_grads) {
+  return std::any_of(
+      out_grads.begin(), out_grads.end(),
+      [](const std::shared_ptr<TensorArg>& tensor_arg) { return !tensor_arg->Empty(); });
+}
+
+Maybe<void> InitEmptyTensorArgs2ZerosTensor(
+    const TensorTuple& outputs, const std::vector<std::shared_ptr<TensorArg>>& out_grads) {
+  for (int i = 0; i < out_grads.size(); i++) {
+    if (out_grads[i]->Empty()) {
```

out_grads.at(i)
任何时候都谨慎使用中括号，因为它的find or create行为不符合fail fast的原则。